libnvme/src/controller.rs

```diff
@@ -0,0 +1,294 @@
+    DontBlock = 1 << 0,
+}
+
+pub enum TryLockResult<L, T, E> {
+    Ok(L),
+    Locked(T),
+    Err(E),
+}
+
+pub struct Controller<'handle> {
+    pub(crate) inner: *mut nvme_ctrl_t,
+    _nvme: &'handle Nvme,
+}
+
+impl<'handle> Controller<'handle> {
+    pub fn get_info(&self) -> Result<ControllerInfo, NvmeError> {
+        let mut ctrl_info: *mut nvme_ctrl_info_t = std::ptr::null_mut();
+        match { unsafe { nvme_ctrl_info_snap(self.inner, &mut ctrl_info) } } {
+            true => Ok(unsafe { ControllerInfo::from_raw(ctrl_info) }),
+            false => {
+                Err(self.fatal_context("failed to get controller snapshot"))
+            }
+        }
+    }
+
+    fn lock_impl(
+        self,
+        level: ControllerLockLevel,
+        flags: ControllerLockFlags,
+    ) -> Result<LockedController<'handle>, (Self, NvmeError)> {
+        match unsafe { nvme_ctrl_lock(self.inner, level as u32, flags as u32) }
+        {
+            true => Ok(LockedController { controller: Some(self) }),
+            false => {
+                let error =
+                    self.fatal_context("failed to grab nvme controller lock");
+                Err((self, error))
+            }
+        }
+    }
+
+    pub fn read_lock(self) -> Result<LockedController<'handle>, NvmeError> {
+        match self
+            .lock_impl(ControllerLockLevel::Read, ControllerLockFlags::Block)
+        {
+            Ok(l) => Ok(l),
+            Err((_, e)) => Err(e),
+        }
+    }
+
+    pub fn write_lock(self) -> Result<LockedController<'handle>, NvmeError> {
+        match self
+            .lock_impl(ControllerLockLevel::Write, ControllerLockFlags::Block)
+        {
+            Ok(l) => Ok(l),
+            Err((_, e)) => Err(e),
+        }
+    }
+
+    pub fn try_read_lock(
+        self,
+    ) -> TryLockResult<LockedController<'handle>, Self, NvmeError> {
+        match self.lock_impl(
+            ControllerLockLevel::Read,
+            ControllerLockFlags::DontBlock,
+        ) {
+            Ok(lock) => TryLockResult::Ok(lock),
+            Err((c, e)) => match e {
+                NvmeError::LockWouldBlock(_) => TryLockResult::Locked(c),
+                e => TryLockResult::Err(e),
+            },
+        }
+    }
+
+    pub fn try_write_lock(
+        self,
+    ) -> TryLockResult<LockedController<'handle>, Self, NvmeError> {
+        match self.lock_impl(
+            ControllerLockLevel::Write,
+            ControllerLockFlags::DontBlock,
+        ) {
+            Ok(lock) => TryLockResult::Ok(lock),
+            Err((c, e)) => match e {
+                NvmeError::LockWouldBlock(_) => TryLockResult::Locked(c),
+                e => TryLockResult::Err(e),
+            },
+        }
+    }
+
+    pub fn namespace_discovery(
+        &self,
+        level: NamespaceDiscoveryLevel,
+    ) -> Result<NamespaceDiscovery, NvmeError> {
+        NamespaceDiscovery::new(self, level)
+    }
+}
+
+impl<'handle> Drop for Controller<'handle> {
```

Comment:
Tiny style nit: if the `impl` block never needs to reference the lifetime, you can elide it here similarly to function signatures:

```suggestion
impl Drop for Controller<'_> {
```